### PR TITLE
Create a new Ginkgo Executor

### PR DESCRIPTION
**Pull request by jdborneman-terminus**
## Pull request description 

Creates a new Ginkgo Test Executor for TestKube.

## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [ ] tested locally
- [ ] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

## Breaking changes

-

## Changes

-

## Fixes

-

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,77 +1,61 @@
...
-# Welcome to TestKube Template Executor
-
-TestKube Template Executor is a test executor skeleton for [TestKube](https://testkube.io).  
-You can use it as basic building blocks for creating a new executor.
-
-# What is an Executor?
-
-Executor is nothing more than a program wrapped into Docker container which gets JSON (testube.Execution) OpenAPI based document as an input and returns a stream of JSON output lines (testkube.ExecutorOutput), where each output line is simply wrapped in this JSON, similar to the structured logging idea. 
-
+Kubetest Ginkgo Executor is test executor for [testkube](https://testkube.io)
```
Comment (jfermi) [inline README.md:5]:
```suggestion
The Kubetest Ginkgo Executor is a test executor for [testkube](https://testkube.io).
```

File: README.md [hunk 2]
```diff
@@ -1,77 +1,61 @@
...
-  name: postman-executor
-  namespace: testkube
-spec:
-  executor_type: job
-  image: kubeshop/testkube-template-executor:0.0.1
-  types:
-  - example/test
-```
-
+Please follow to main testkube repository for reporting any [issues](https://github.com/kubeshop/testkube/issues) or [discussions](https://github.com/kubeshop/testkube/discussions)
```
Comment (jfermi) [inline README.md:9]:
```suggestion
Please visit the main Testkube repository for reporting any [issues](https://github.com/kubeshop/testkube/issues) or [discussions](https://github.com/kubeshop/testkube/discussions).
```

File: README.md [hunk 3]
```diff
@@ -1,77 +1,61 @@
...
-![Release](https://img.shields.io/github/v/release/kubeshop/testkube) [![Releases](https://img.shields.io/github/downloads/kubeshop/testkube/total.svg)](https://github.com/kubeshop/testkube/tags?label=Downloads) ![Go version](https://img.shields.io/github/go-mod/go-version/kubeshop/testkube)
+- TODO add architecture diagrams
 
-![Docker builds](https://img.shields.io/docker/automated/kubeshop/testkube-api-server) ![Code build](https://img.shields.io/github/workflow/status/kubeshop/testkube/Code%20build%20and%20checks) ![Release date](https://img.shields.io/github/release-date/kubeshop/testkube)
+## API 
 
-![Twitter](https://img.shields.io/twitter/follow/thekubeshop?style=social) ![Discord](https://img.shields.io/discord/884464549347074049)
- #### [Documentation](https://kubeshop.github.io/testkube) | [Discord](https://discord.gg/hfq44wtR6Q) 
\ No newline at end of file
+Cypress executor implements [testkube OpenAPI for executors](https://kubeshop.github.io/testkube/openapi/#operations-tag-executor) (look at executor tag)
```
Comment (jfermi) [inline README.md:61]:
```suggestion
Cypress executor implements [testkube OpenAPI for executors](https://kubeshop.github.io/testkube/openapi/#operations-tag-executor) (look at executor tag).
```

**Review by jfermi — COMMENTED** [review]
A few editing suggestions.

**Comment by jdborneman-terminus** [comment]
> A few editing suggestions.

Thanks! I'll take a look between my meetings on Endless Meeting Thursday 😁

**Comment by jdborneman-terminus** [comment]
> looks good, great work! I miss 2 things:
> 
> 1. examples folder - with success and failure examples
> 2. ideally - success and failure unit tests, that should be run if ginko is installed, skip otherwise

@vsukhin 
* examples folder: Do you mean example ginkgo test project with successes and failures examples?
* unit tests: have a good example in say.. .cypress or something that accomplishes what you want to see that I can copy off of a bit?

Thanks!

**Comment by vsukhin** [comment]
1. yes
2. yeah, anything you like, it can be dead simple, for example, we just check in curl executor, that https://reqbin.com/echo/get/json returns "success"

**Comment by jdborneman-terminus** [comment]
> 

@vsukhin okay cool. Couple more questions though

1. if I include example ginkgo tests in the project that both pass and fail, the `go-build.yml` workflow will fail the Test step as it will end up running them. How, with `go test -v ./...` can I tell it to just run the tests that are in `pkg/runner/runner_test.go`? I tried `go test -v pkg/runner/runner_test.go` but just got errors that all the funcs from runner.go are undefined.
```
% go test -v pkg/runner/runner_test.go
# command-line-arguments [command-line-arguments.test]
pkg/runner/runner_test.go:12:20: undefined: InitializeGinkgoParams
pkg/runner/runner_test.go:24:20: undefined: InitializeGinkgoParams
pkg/runner/runner_test.go:39:19: undefined: FindGinkgoParams
pkg/runner/runner_test.go:45:20: undefined: InitializeGinkgoParams
pkg/runner/runner_test.go:46:15: undefined: BuildGinkgoArgs
pkg/runner/runner_test.go:78:19: undefined: BuildGinkgoPassThroughFlags
FAIL    command-line-arguments [build failed]
```
2. Would the runner then need to support like... a local Dir source for the tests? Right now it just supports a github repo reference.

THanks

**Comment by jdborneman-terminus** [comment]
Ah just looking at 
https://github.com/kubeshop/testkube-executor-cypress/blob/main/pkg/runner/cypress_test.go

Looks like it references itself as a repo and tests it that way. Okay. QUestion 1 still stands how to include some more ginkgo tests but NOT run them with `go test`

**Comment by jdborneman-terminus** [comment]
Hm, this is fighting me. I'm trying to add a test of GinkgoRunner that runs tests from a repo.

```
t.Run("GinkgoRunner should run tests from a repo", func(t *testing.T) {
		runner, err := NewGinkgoRunner()
		if err != nil {
			t.Fail()
		}
		repoURI := "https://github.com/jdborneman-terminus/testkube-executor-ginkgo.git"
		result, err := runner.Run(testkube.Execution{
			Content: &testkube.TestContent{
				Type_: string(testkube.TestContentTypeGitDir),
				Repository: &testkube.Repository{
					Type_:  "git",
					Uri:    repoURI,
					Branch: "main",
					Path:   "examples",
				},
			},
		})

		fmt.Printf("RESULT: %+v\n", result)
		fmt.Printf("ERROR:  %+v\n", err)
		t.Fail()
	})
```

BUt when I try to run it I get all sorts of runtime error invalid memory address or nil pointer dereference. 
```
% go test -v pkg/runner/runner_test.go pkg/runner/runner.go
{"type":"event","content":"Initialized Ginkgo Parameters. Count: [25]"}
=== RUN   TestRun
=== RUN   TestRun/GinkgoRunner_should_run_tests_from_a_repo
--- FAIL: TestRun (0.00s)
    --- FAIL: TestRun/GinkgoRunner_should_run_tests_from_a_repo (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x12a0551]
```

I copied this pattern from teh Cypress test here: https://github.com/kubeshop/testkube-executor-cypress/blob/main/pkg/runner/cypress_test.go

But I noticed that it's skipped and not run. Is there an example somewhere that DOES run?  I looked at executor_curl but didn't find an example of this sort of test.

@vsukhin 

Thanks!

**Comment by vsukhin** [comment]
This is why we need unit tests - to catch missing stuff ) Probably, something is not initialised, like Variables map or soemthing similar. Find the place where panic is generated

**Comment by jdborneman-terminus** [comment]
Added two new tests, @vsukhin.

1. Tests that GinkgoRunner should run tests from a repo/pkg that pass
2. Tests that GinkgoRunner should run tests from a repo/pkg that fail

One note is that the repoURI will need changed in those tests AFTER things are merged to the kubeshop/testkube-executor-ginkgo repo. Right now they reference my fork.

**Comment by vsukhin** [comment]
Sure, I saw it. You can create one line pr then

**Review by vsukhin — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -1,77 +1,61 @@
...
+* `GinkgoCover`, default: `""`, usage: `--cover`
+* `GinkgoCoverProfile`, default: `""`, usage: `--coverprofile cover.profile`
+* `GinkgoRace`, default: `""`, usage: `--race`
+* `GinkgoTrace`, default: `"--trace"`
+* `GinkgoJsonReport`, default: `""`, usage: `--json-report report.json`
+* `GinkgoJunitReport`, default: `"--junit-report report.xml"`
+* `GinkgoTeamCityReport`, default: `""`, usage: `--teamcity-report report.teamcity`
+
+### Pass-through args to Ginkgo:
+Add `--args '--base-url=example.com --some-arg=value'` to `kubectl testkuve run test` command.
```
Comment (nicufk) [inline README.md:47]:
```suggestion
Add `--args '--base-url=example.com --some-arg=value'` to `kubectl testkube run test` command.
```

**Review by nicufk — APPROVED** [review]
LGTM

**Comment by jdborneman-terminus** [comment]
@jfermi @nicufk @vsukhin I made the necessary changes to support 1.3.17 with testkueb create with commit id. If you want to re-review.

But this works locally for me perfectly. Just what we'll need. I think it's ready for merge/release whenever you all say it is.

**Pull request merged**